core/pta/device.c

```diff
@@ -21,18 +21,20 @@
 #define PTA_NAME "device.pta"
 
 static void add_ta(uint32_t flags, const TEE_UUID *uuid, uint8_t *buf,
-		   uint32_t blen, uint32_t *pos)
+		   uint32_t blen, uint32_t *pos, uint32_t rflags)
 {
-	if (flags & TA_FLAG_DEVICE_ENUM) {
-		if (*pos + sizeof(*uuid) <= blen)
-			tee_uuid_to_octets(buf + *pos, uuid);
+	if (!(flags & rflags & TA_FLAG_DEVICE_ENUM_MASK))
+		return;
```

Comment:
I think we need an additional check here to not report a TA which sets both flags `TA_FLAG_DEVICE_ENUM` and `TA_FLAG_DEVICE_ENUM_SUPP`. As it could lead to double registration of same TA as device which seems to be incorrect. And perhaps we could put a warning message to show that its an incorrect combination.

Reply:
I think if both flags are specified then TA /driver is written in that way that double initialization is expected. However it might be not easy to think up such use case.

Reply:
It isn't like the same device will be initialized twice but it will be like two separate devices registered for single TA and each device probed at once which doesn't sounds correct.

While at it, haven't you faced following error while enabling supplicant enumeration:

```
[    2.330622] sysfs: cannot create duplicate filename '/devices/optee-clnt0'
[    2.331060] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.7.0-rc5-01053-g43a7eab-dirty #4
[    2.331268] Hardware name: QEMU QEMU Virtual Machine, BIOS 0.0.0 02/06/2015
[    2.331585] Call trace:
<snip>
```
As I think you need to update logic in Linux TEE device registration here [1] as well to give different names to devices enumerated without supplicant dependency and with supplicant dependency.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/optee/device.c#n78

Reply:
I didn't setup 2 flags in one time. But yes, there is bunch of places where linux drivers needs to be changes (like you pointed to sysfs entries). I will add check for now and if there will be use case then we can rethink how to support 2 flags together.